File: java/DivideAndConquer/binarySearch.java
public class binarySearch {
    public static void main(String[] args) {
        int[] arr = {1, 2, 3, 4, 5, 6, 7, 8, 9, 10};
        System.out.println(binarySearch(arr, 10));
    }

    public static boolean binarySearch(int[] arr, int target) {

        // check if theres an ampty array given
        if(arr.length == 0) {
            return false;
        }

        // here we want an the index of the first and last elements in the array
        int lowend = 0;
        int highend = arr.length-1;   

        // we'll keep cutting the problem size in half until either the highend reaches 0
        // or until we find our target
        while(lowend <= highend) {
            int midpoint = highend - lowend / 2;

            // here, our guess was larger than the target
            // therefore we eliminate everything on the right side of the midpoint
            if(arr[midpoint] > target) {
                highend = midpoint-1;
            }

            // here our guess is smaller than the target
            // therefore we  elimiate the left side of the midpoint
            else if(arr[midpoint] < target) {
                lowend = midpoint - 1;

            // we have found the target
            } else {
                return true;
            }
        }
        return false;
    }
}

File: swift/DivideAndConquer/binarySearch.swift
import Foundation

public class BinarySearch {
    static func binarySearch(for arr: Array<Int>, target: Int) -> Bool {
        var upperBound = arr.count-1
        var lowerBound = 0
        while lowerBound <= upperBound {
            let midPoint = upperBound - lowerBound / 2
            print(midPoint)
            if arr[midPoint] > target {
                upperBound = midPoint - 1
            } else if arr[midPoint] < target {
                lowerBound = midPoint + 1
            } else {
                return true
            }
        }
        return false
    }
}

let arr = [0, 1, 2, 3, 4, 5, 6, 7, 8, 9, 10, 11, 12, 13, 14, 15, 16]
let target = 15
let result = BinarySearch.binarySearch(for: arr, target: target)
print(result)
